### PR TITLE
feat: add collection view for Web3 tab

### LINKED ISSUE #4751: [Demand] View NFT tokens in collections
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Web3)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Add a switch as "View in collections" in the Web3 tab. If the user turns it on, the NFT tokens will be listed in collection groups.

### PR DESCRIPTION
**Pull request by yanzhihong23**
closes #4751

image

image

### CODE REVIEW

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 1]
```diff
@@ -115,6 +139,15 @@ function CollectibleListUI(props: CollectibleListUIProps) {
             </Box>
         )
 
+    const collections: ERC721TokenDetailed[][] = []
+    for (let i = 0; i < collectibles.length; i = i + 1) {
+        if (i && collectibles[i].contractDetailed.address === collectibles[i - 1].contractDetailed.address) {
```
Comment (guanbinrui) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:144]: Use `isSameAddress` to compare two addresses.

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 2]
```diff
@@ -115,6 +139,15 @@ function CollectibleListUI(props: CollectibleListUIProps) {
             </Box>
         )
 
+    const collections: ERC721TokenDetailed[][] = []
```
Comment (guanbinrui) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:142]: Why not a hook for this?

**Review by guanbinrui — DISMISSED** [review]

**Comment by wenluomask** [comment]
1. Some NFTs didn't show the correct collection, you may fetch the worng session of openses api;
截屏2021-11-11 下午5 22 43

2. Some NFTs belong to same collection, but it display in different place;(ex. ENS)
截屏2021-11-11 下午5 31 49

3. Since seperate by collections, you need to sort out all of NFTs owned by user, not only sort by pages;

File: packages/dashboard/src/web3/context.ts [hunk 3]
```diff
@@ -92,6 +92,7 @@ export const Web3Context: Web3ProviderType = {
 
     getAssetsList: PluginServices.Wallet.getAssetsList,
     getAssetsListNFT: PluginServices.Wallet.getAssetsListNFT,
+    getCollectionsNFT: PluginServices.Wallet.getCollectionsNFT,
```
Comment (guanbinrui) [inline packages/dashboard/src/web3/context.ts:95]: IMO it duplicates with `getAssetsListNFT`.

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 4]
```diff
@@ -183,18 +189,28 @@ export function CollectibleListAddress(props: CollectibleListAddressProps) {
...
 
     useEffect(() => {
-        const tab = searchProfileTabSelector().evaluate()
-        if (!tab) return
-        tab.scrollIntoView()
-    }, [page])
+        if (collectibles.length) {
+            setRendCollectibles([...rendCollectibles, ...collectibles])
+            if (hasNextPage) {
+                setTimeout(() => {
```
Comment (guanbinrui) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:204]: Clean the timer.

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 5]
```diff
@@ -205,11 +221,63 @@ export function CollectibleListAddress(props: CollectibleListAddressProps) {
...
+
+    useUpdateEffect(() => {
+        setPage(0)
+    }, [provider, address])
+
+    useEffect(() => {
+        if (collections.length) {
+            setRendCollections([...rendCollections, ...collections])
+            if (hasNextPage) {
+                setTimeout(() => {
```
Comment (guanbinrui) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:248]: Clean the timer.

**Comment by wenluomask** [comment]
Checked.

**Review by guanbinrui — DISMISSED** [review]

**Pull request merged**